# Importing Libraries

When working with data analysis and statistical tests, the following libraries are commonly used:

### pandas
- **Purpose**: Used for data manipulation and analysis. It provides powerful data structures (like DataFrames) for reading and processing datasets, such as `.csv` files.
  
### numpy
- **Purpose**: Provides support for numerical operations and array manipulations. It is used to manipulate masks of missing values effectively.

### scipy.stats import chi2
- **Purpose**: Imports the chi-square distribution from the `scipy` library. It is utilized for calculating the chi-square statistic and the corresponding p-value in statistical tests.


In [1]:
import pandas as pd
import numpy as np
from scipy.stats import chi2

### Loading a CSV File into a Pandas DataFrame

Using `pd.read_csv`, we can easily load a CSV file into a pandas DataFrame. This function reads the CSV file and creates a DataFrame that allows for efficient data manipulation and analysis.

**Key Points**:
- **Function**: `pd.read_csv(filepath)`
- **Parameters**:
  - `filepath`: The path to the CSV file you want to load.
  - Additional parameters can be used to customize the loading process (e.g., `delimiter`, `header`, `na_values`).

This method is essential for importing data from external sources for further analysis.


In [2]:
df = pd.read_csv('startup_funding.csv')

### Retrieving Null Value Counts in a DataFrame

Using `df.isnull().sum()`, we can easily retrieve the number of null values present in each column of a pandas DataFrame. 

**Key Points**:
- **Function**: `df.isnull().sum()`
  - `df.isnull()`: Creates a boolean DataFrame indicating `True` for null values and `False` for non-null values.
  - `.sum()`: Sums up the `True` values (which are treated as 1) for each column, resulting in the total count of null values.

This method is useful for assessing the completeness of data in each column before performing further analysis or data cleaning.


In [3]:
df.isnull().sum()

SNo                    0
Date                   0
StartupName            0
IndustryVertical     171
SubVertical          936
CityLocation         179
InvestorsName          8
InvestmentType         1
AmountInUSD          847
Remarks             1953
dtype: int64

### Displaying the First 5 Rows of a Dataset

Using `df.head()`, we can easily display the first five rows of a pandas DataFrame. 

**Key Points**:
- **Function**: `df.head()`
  - This function returns the first five rows of the DataFrame by default, allowing for a quick preview of the data.
  - You can also specify the number of rows to display by passing an integer as an argument (e.g., `df.head(10)` for the first ten rows).

This method is helpful for quickly inspecting the structure and content of the dataset.


In [4]:
df.head()

SNo        Date   StartupName   IndustryVertical  \
0    0  01/08/2017      TouchKin         Technology   
1    1  02/08/2017       Ethinos         Technology   
2    2  02/08/2017  Leverage Edu  Consumer Internet   
3    3  02/08/2017          Zepo  Consumer Internet   
4    4  02/08/2017  Click2Clinic  Consumer Internet   

                                     SubVertical CityLocation  \
0                       Predictive Care Platform    Bangalore   
1                       Digital Marketing Agency       Mumbai   
2  Online platform for Higher Education Services    New Delhi   
3                         DIY Ecommerce platform       Mumbai   
4                  healthcare service aggregator    Hyderabad   

                                       InvestorsName  InvestmentType  \
0                                        Kae Capital  Private Equity   
1                         Triton Investment Advisors  Private Equity   
2  Kashyap Deorah, Anand Sankeshwar, Deepak Jain,...    Seed Funding   
3  Kunal Shah, LetsVenture, Anupam Mittal, Hetal ...    Seed Funding   
4                     Narottam Thudi, Shireesh Palle    Seed Funding   

  AmountInUSD Remarks  
0   1,300,000     NaN  
1         NaN     NaN  
2         NaN     NaN  
3     500,000     NaN  
4     850,000     NaN

The `df.info()` method in pandas provides a concise summary of a DataFrame, giving essential information about its structure and contents. It is especially useful for quickly understanding the dataset you're working with.

### Key Details Provided by `df.info()`:
1. **Number of Rows and Columns**:
   - It shows the total number of rows and columns in the DataFrame, helping you understand the size of your dataset.

2. **Column Names and Data Types**:
   - Lists all the columns in the DataFrame along with their respective data types (e.g., `int64`, `float64`, `object`, etc.). This helps in identifying potential issues with data types, such as a numeric column being stored as `object`.

3. **Non-Null Count**:
   - Displays the count of non-null values for each column. This allows you to see how much data is missing in each column without manually checking for null values.

4. **Memory Usage**:
   - Provides the memory usage of the DataFrame, which is helpful for optimizing the DataFrame’s size when working with large datasets.



In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2372 entries, 0 to 2371
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   SNo               2372 non-null   int64 
 1   Date              2372 non-null   object
 2   StartupName       2372 non-null   object
 3   IndustryVertical  2201 non-null   object
 4   SubVertical       1436 non-null   object
 5   CityLocation      2193 non-null   object
 6   InvestorsName     2364 non-null   object
 7   InvestmentType    2371 non-null   object
 8   AmountInUSD       1525 non-null   object
 9   Remarks           419 non-null    object
dtypes: int64(1), object(9)
memory usage: 185.4+ KB


### Converting `AmountInUSD` to Integer Type

To convert the column named `AmountInUSD` to an integer type, we need to follow these steps:

1. **Replace Commas**: Since the values are stored as objects and may contain commas (e.g., "1,000"), we first need to replace commas with empty strings.
2. **Fill NaN Values**: Next, we fill any NaN values with `0` to avoid errors during the conversion.
3. **Convert to Integer**: Finally, we use the `astype()` function to convert the cleaned column to an integer type.

**Key Steps**:
- Use `df['AmountInUSD'].str.replace(",", "")` to remove commas.
- Use `df['AmountInUSD'].fillna(0, inplace=True)` to fill NaN values.
- Use `df['AmountInUSD'] = df['AmountInUSD'].astype(int)` to convert the column to integers.

This process ensures that the `AmountInUSD` column is properly formatted for numerical analysis.


In [6]:
df['AmountInUSD']

0       1,300,000
1             NaN
2             NaN
3         500,000
4         850,000
          ...    
2367    4,500,000
2368      825,000
2369    1,500,000
2370          NaN
2371      140,000
Name: AmountInUSD, Length: 2372, dtype: object

In [7]:
df['AmountInUSD'] = df['AmountInUSD'].str.replace(',','')

In [8]:
df['AmountInUSD']

0       1300000
1           NaN
2           NaN
3        500000
4        850000
         ...   
2367    4500000
2368     825000
2369    1500000
2370        NaN
2371     140000
Name: AmountInUSD, Length: 2372, dtype: object

In [9]:
df['AmountInUSD'] =df['AmountInUSD'].fillna(0).astype(int)

In [10]:
df['AmountInUSD']

0       1300000
1             0
2             0
3        500000
4        850000
         ...   
2367    4500000
2368     825000
2369    1500000
2370          0
2371     140000
Name: AmountInUSD, Length: 2372, dtype: int32

### `df.isnull().sum()` Usage

The `df.isnull().sum()` method in pandas is used to quickly identify and count the number of missing (`NaN`) values in each column of a DataFrame. This is particularly useful for checking data quality and understanding how much data is missing from each feature.

#### How It Works:
- **`df.isnull()`**: This returns a DataFrame of the same shape as `df`, with `True` where values are `NaN` and `False` otherwise.
- **`sum()`**: When applied after `isnull()`, it counts the number of `True` values (i.e., missing values) for each column.



In [11]:
df.isnull().sum()

SNo                    0
Date                   0
StartupName            0
IndustryVertical     171
SubVertical          936
CityLocation         179
InvestorsName          8
InvestmentType         1
AmountInUSD            0
Remarks             1953
dtype: int64

In [12]:
len(df)

2372

In [13]:
null_count = df.isnull().sum()

### `null_count_df = null_count.reset_index()` Usage

The line `null_count_df = null_count.reset_index()` is used to convert a Series (in this case, `null_count`, which contains the count of missing values in each column) into a DataFrame, while also resetting the index. This operation is useful when you want to convert a summary of missing values into a DataFrame format for further manipulation or visualization.

#### How It Works:
1. **`null_count`**: This is typically a pandas Series (for example, the result of `df.isnull().sum()`) where the index contains column names, and the values represent the count of missing values in those columns.
2. **`reset_index()`**: This method turns the index into a regular column, converting the Series into a DataFrame. The original index becomes the first column, and the data becomes the second column.


In [14]:
null_count_df = null_count.reset_index()

In [15]:
null_count_df

index     0
0               SNo     0
1              Date     0
2       StartupName     0
3  IndustryVertical   171
4       SubVertical   936
5      CityLocation   179
6     InvestorsName     8
7    InvestmentType     1
8       AmountInUSD     0
9           Remarks  1953

In [16]:
null_count_df['index'=='3'][3]

171

### Importance of Missing Value Percentage in Terms of MCAR (Missing Completely at Random)

Understanding the percentage of missing values in a dataset is crucial, particularly when considering the assumption of **MCAR (Missing Completely at Random)**. If data is MCAR, the missingness occurs independently of both observed and unobserved data, meaning it does not depend on the actual values of any variables.

#### Why is Missing Percentage Important?

1. **Bias in Analysis**:
   - High percentages of missing values can lead to biased analyses, especially if the data is not MCAR. Non-MCAR missingness may depend on other variables or values, potentially distorting results.

2. **MCAR Assumption**:
   - If data is truly missing at random (MCAR), common techniques like listwise deletion or imputation generally won't introduce bias. However, the percentage of missing data matters—large amounts of missingness reduce the effective sample size and may affect model accuracy.

3. **Data Quality**:
   - The percentage of missing data gives a measure of the dataset's quality and reliability. If key columns have high missing percentages, additional investigation may be necessary to determine the cause and the best method for handling it.

Understanding the percentage of missing values helps to decide on the right approach for handling them, depending on whether the MCAR assumption holds true.


In [17]:
print(f"% of value of missing in Investment column is {(null_count_df['index'=='3'][7]/len(df))*100}")
print(f"% of value of missing in Remarks column is {(null_count_df['index'=='3'][9]/len(df))*100}")
print(f"% of value of missing in IndustryVerical column is {(null_count_df['index'=='3'][3]/len(df))*100}")
print(f"% of value of missing in subVertical column is {(null_count_df['index'=='3'][4]/len(df))*100}")
print(f"% of value of missing in CityLocation column is {(null_count_df['index'=='3'][5]/len(df))*100}")
print(f"% of value of missing in Investorsname column is {(null_count_df['index'=='3'][6]/len(df))*100}")

% of value of missing in Investment column is 0.042158516020236084
% of value of missing in Remarks column is 82.33558178752108
% of value of missing in IndustryVerical column is 7.209106239460371
% of value of missing in subVertical column is 39.46037099494098
% of value of missing in CityLocation column is 7.546374367622259
% of value of missing in Investorsname column is 0.33726812816188867


### Little's MCAR Test

**Little's MCAR (Missing Completely at Random) Test** is a statistical test used to determine whether missing data in a dataset can be considered MCAR. The test checks if the missingness in the data is unrelated to both observed and unobserved variables, which is the assumption of MCAR.

#### Key Points:
- **Null Hypothesis (H₀)**: The data is MCAR (missing completely at random).
- **Alternative Hypothesis (H₁)**: The data is not MCAR, implying that the missingness is related to other variables (MAR or MNAR).
- **Test Output**: 
  - A p-value is produced. 
  - If the p-value is large (e.g., p > 0.05), we fail to reject H₀, meaning the data can be assumed to be MCAR.
  - If the p-value is small (e.g., p < 0.05), we reject H₀, indicating the missing data is likely not MCAR.

#### Usage:
Little’s MCAR test helps to validate the assumption that data is missing at random, guiding decisions about how to handle missing data. If data is not MCAR, more complex imputation or modeling methods may be needed to address the missingness.

**Limitations**: Little's MCAR test has reduced power with smaller sample sizes or large amounts of missing data, so results should be interpreted cautiously.



### User-Defined Function for Little's MCAR Test

The following user-defined function, `little_mcar_test(data)`, implements a basic version of **Little's MCAR Test** to determine whether missing data is Missing Completely at Random (MCAR).

#### Function Breakdown:
- **Input**: 
  - The function takes a dataset (`data`) as input. 
  - It checks for missing values in each column and creates binary masks where `1` represents a missing value and `0` represents a non-missing value.
  
- **Core Logic**: 
  - For each column with missing data, a mask is created. These masks are concatenated into a binary matrix `r`, where each row represents a sample and each column represents the missing pattern of a variable.
  - The function computes the matrix product `r.T @ r` to summarize the missing patterns across columns.
  - Using this, the **chi-squared statistic** (`chi2_stat`) is calculated as the trace of the matrix.
  
- **Degrees of Freedom**: 
  - The degrees of freedom (`df`) are calculated as `(n - 1) * m`, where `n` is the number of rows in the dataset, and `m` is the number of columns with missing data.

- **Chi-Square Test**: 
  - The chi-squared test statistic and the degrees of freedom are used to compute the p-value (`p_value`), which helps determine if the data is MCAR.

#### Output:
- The function returns a dictionary containing:
  - `chi2_stat`: The chi-squared test statistic.
  - `degree_of_freedom`: Degrees of freedom for the test.
  - `p_value`: The p-value used to assess whether the data is MCAR.

#### Interpretation:
- If the **p-value** is greater than a significance threshold (e.g., p > 0.05), the missing data can be considered **MCAR**.
- If the **p-value** is small (e.g., p < 0.05), we reject the MCAR assumption, implying the missing data may not be random (MAR or MNAR).

This custom function provides a straightforward way to test for the MCAR assumption, aiding in the analysis of missing data.



In [18]:
def little_mcar_test(data):
    n = len(data)
    groups = []
    for col in data.columns:
        mask = data[col].isnull()
        if mask.any():
            groups.append(mask.astype(int).values.reshape(-1,1))
    if len(groups) == 0:
        raise ValueError("No missing data found")
    r = np.concatenate(groups,axis=1)
    group_stats = r.T @ r
    m = len(groups)
    df = (n-1) * m
    chi2_stat = group_stats.trace()
    p_value = chi2.sf(chi2_stat,df)
    return {"chi2_stat":chi2_stat,"degree_of_freedom":df,"p_value":p_value}

In [19]:
result = little_mcar_test(df)
print(f"Chi-square statistic:{result['chi2_stat']}")
print(f"Degrees of freedom:{result['degree_of_freedom']}")
print(f"P-Value:{result['p_value']}")

Chi-square statistic:3248
Degrees of freedom:14226
P-Value:1.0


### Methods for Handling MCAR (Missing Completely at Random) Data

When data is Missing Completely at Random (MCAR), the following methods can be used to handle missing values effectively:

1. **List-wise Deletion**:
   - In this method, any record with missing data in any of its variables is completely deleted from the dataset.
   - This method is best suited when the percentage of missing data is very small (e.g., less than 2%).
   - **Drawback**: It can lead to a significant reduction in the dataset size, especially when multiple variables have missing values.

2. **Pair-wise Deletion**:
   - Unlike list-wise deletion, pair-wise deletion only removes data when the specific pair of variables used in a statistical method has missing values.
   - It works similarly to how a correlation matrix is calculated, where missing values are ignored between the two variables being analyzed.
   - **Advantage**: This approach minimizes data loss as only the incomplete pairs are excluded from analysis, rather than entire records.
   - **Use Case**: Pair-wise deletion is a better option when many variables in the dataset have missing values but different variables are missing across different cases.

   **Key Difference**: 
   - List-wise deletion reduces the number of complete cases by removing entire rows, while pair-wise deletion adjusts the number of cases depending on the specific variables involved in the analysis.

3. **Mean, Median & Mode Imputation**:
   - This method involves replacing missing values with the mean, median, or mode of the respective variable.
     - **Mean**: Best for continuous numerical data when the data is normally distributed.
     - **Median**: Suitable when the data has skewness, as the median is less affected by outliers.
     - **Mode**: Commonly used for categorical variables.
   - **Advantage**: This is a simple and effective method that retains all records.
   - **Drawback**: It may introduce bias into the dataset, especially if the underlying data distribution is complex.

Each method has its trade-offs, and the choice depends on the amount and pattern of missing data, as well as the dataset's structure and analysis requirements.


### Columns to Apply List-wise Deletion

- **Investment** (0.04% missing) and **Investorsname** (0.34% missing): 
  - Suitable for list-wise deletion since the missing percentage is very small, and it will not significantly impact the dataset size.
  
- **Remarks** (82.34%) and **SubVertical** (39.46%): 
  - Avoid list-wise deletion due to the high percentage of missing data. Consider using imputation methods instead.

- **IndustryVertical** (7.21%) and **CityLocation** (7.55%): 
  - List-wise deletion could be applied, but since the missing percentage is moderate, imputation may be a better choice to minimize data loss.


In [20]:
df2 =df.copy()

In [21]:
df2_cleaned = df2.dropna(subset=['InvestorsName','InvestmentType'])

In [22]:
df2_cleaned.isnull().sum()

SNo                    0
Date                   0
StartupName            0
IndustryVertical     171
SubVertical          930
CityLocation         179
InvestorsName          0
InvestmentType         0
AmountInUSD            0
Remarks             1945
dtype: int64

### Imputation of Categorical Columns Using Mode

For categorical columns with missing values, such as **Remarks**, **SubVertical**, **CityLocation**, and **IndustryVertical**, imputation using the **mode** is a suitable approach. The mode represents the most frequently occurring value in a dataset, making it a logical choice for filling in missing categorical data.

#### Steps for Imputation:
1. **Identify Categorical Columns**: Determine which columns are categorical and contain missing values.
2. **Calculate Mode**: Use the mode function to find the most frequent value in each categorical column.
3. **Impute Missing Values**: Replace missing values in these columns with their respective mode values.

In [23]:
df3 = df.copy()

In [24]:
df3.isnull().sum()

SNo                    0
Date                   0
StartupName            0
IndustryVertical     171
SubVertical          936
CityLocation         179
InvestorsName          8
InvestmentType         1
AmountInUSD            0
Remarks             1953
dtype: int64

In [25]:
df3['Remarks']

0                                     NaN
1                                     NaN
2                                     NaN
3                                     NaN
4                                     NaN
                      ...                
2367                                  NaN
2368                  Govt backed VC Fund
2369                                  NaN
2370    Strategic Funding, Minority stake
2371                                  NaN
Name: Remarks, Length: 2372, dtype: object

In [28]:
mode_value = df3['Remarks'].mode()[0]

In [29]:
df3['Remarks'] = df3['Remarks'].fillna(mode_value)

In [30]:
df3.isnull().sum()

SNo                   0
Date                  0
StartupName           0
IndustryVertical    171
SubVertical         936
CityLocation        179
InvestorsName         8
InvestmentType        1
AmountInUSD           0
Remarks               0
dtype: int64

In [31]:
df4 = df.copy()

In [32]:
df4['SubVertical']

0                            Predictive Care Platform
1                            Digital Marketing Agency
2       Online platform for Higher Education Services
3                              DIY Ecommerce platform
4                       healthcare service aggregator
                            ...                      
2367                                              NaN
2368                                              NaN
2369                                              NaN
2370                                              NaN
2371                                              NaN
Name: SubVertical, Length: 2372, dtype: object

In [34]:
mode_value_sub = df4['SubVertical'].mode()[0]

In [35]:
mode_value_sub

'Online Pharmacy'

In [36]:
df4['SubVertical'] = df4['SubVertical'].fillna(mode_value_sub)

In [37]:
df4.isnull().sum()

SNo                    0
Date                   0
StartupName            0
IndustryVertical     171
SubVertical            0
CityLocation         179
InvestorsName          8
InvestmentType         1
AmountInUSD            0
Remarks             1953
dtype: int64

In [38]:
df5 = df.copy()

In [39]:
df5['CityLocation']

0       Bangalore
1          Mumbai
2       New Delhi
3          Mumbai
4       Hyderabad
          ...    
2367          NaN
2368          NaN
2369          NaN
2370          NaN
2371          NaN
Name: CityLocation, Length: 2372, dtype: object

In [40]:
mode_value_city = df5['CityLocation'].mode()[0]

In [41]:
mode_value_city

'Bangalore'

In [42]:
df5['CityLocation'] = df5['CityLocation'].fillna(mode_value_city)

In [43]:
df5.isnull().sum()

SNo                    0
Date                   0
StartupName            0
IndustryVertical     171
SubVertical          936
CityLocation           0
InvestorsName          8
InvestmentType         1
AmountInUSD            0
Remarks             1953
dtype: int64

In [44]:
df6 = df.copy()

In [45]:
df6.isnull().sum()

SNo                    0
Date                   0
StartupName            0
IndustryVertical     171
SubVertical          936
CityLocation         179
InvestorsName          8
InvestmentType         1
AmountInUSD            0
Remarks             1953
dtype: int64

In [46]:
df6['IndustryVertical']

0              Technology
1              Technology
2       Consumer Internet
3       Consumer Internet
4       Consumer Internet
              ...        
2367                  NaN
2368                  NaN
2369                  NaN
2370                  NaN
2371                  NaN
Name: IndustryVertical, Length: 2372, dtype: object

In [47]:
mode_value_industry_verti = df6['IndustryVertical'].mode()[0]

In [48]:
mode_value_industry_verti

'Consumer Internet'

In [49]:
df6['IndustryVertical'] = df6['IndustryVertical'].fillna(mode_value_industry_verti)

In [50]:
df6.isnull().sum()

SNo                    0
Date                   0
StartupName            0
IndustryVertical       0
SubVertical          936
CityLocation         179
InvestorsName          8
InvestmentType         1
AmountInUSD            0
Remarks             1953
dtype: int64